File: bill_tracker/bill_tracker/report/monthly_invoice_matrix/monthly_invoice_matrix.py
import frappe
from datetime import datetime


def execute(filters=None):
    columns = [
        {
            "label": "Customer",
            "fieldname": "customer_id",
            "fieldtype": "Link",
            "options": "Customers",
            "width": 180,
        },
    ]

    raw_invoices = frappe.get_all("Customer Invoice", fields=["invoice_date"], distinct=True)

    def parse_month(month):
        try:
            return datetime.strptime(month, "%b-%Y")
        except (TypeError, ValueError):
            return datetime.min

    months = sorted(
        {invoice.invoice_date.strftime("%b-%Y") for invoice in raw_invoices if invoice.invoice_date},
        key=parse_month,
    )

    for month in months:
        month_key = month.replace("-", "_").lower()
        columns.append({
            "label": f"<b>{month}</b>",
            "fieldname": month_key,
            "fieldtype": "Data",
            "width": 150,
        })

    customers = frappe.get_all("Customers", fields=["name", "customer_name"])
    invoices = frappe.get_all(
        "Customer Invoice",
        fields=[
            "name",
            "customer_id",
            "invoice_date",
            "invoice_total_amount",
            "invoice_paid_amount",
            "invoice_outstanding_amount",
            "status",
        ],
    )

    invoice_map = {
        (invoice.customer_id, invoice.invoice_date.strftime("%b-%Y")): invoice
        for invoice in invoices
        if invoice.customer_id and invoice.invoice_date
    }

    data = []
    for customer in customers:
        row = {
            "customer_id": customer.name,
            "customer_name": customer.customer_name,
            "is_summary": False,
        }
        has_pending = False
        has_entries = False

        for month in months:
            month_key = month.replace("-", "_").lower()
            invoice = invoice_map.get((customer.name, month))
            if invoice:
                has_entries = True
                row[f"doc_{month_key}"] = invoice.name
                row[month_key] = invoice.invoice_total_amount or 0
                row[f"date_{month_key}"] = invoice.invoice_date
                row[f"status_{month_key}"] = invoice.status or "Draft"
                row[f"paid_{month_key}"] = invoice.invoice_paid_amount or 0
                row[f"outstanding_{month_key}"] = invoice.invoice_outstanding_amount or 0

                if invoice.status != "Paid":
                    has_pending = True
            else:
                row[f"status_{month_key}"] = "-"

        if has_pending:
            row["overall_status"] = "Pending"
        elif has_entries:
            row["overall_status"] = "Paid"
        else:
            row["overall_status"] = "None"

        data.append(row)

    return columns, data

File: bill_tracker/bill_tracker/report/monthly_invoice_matrix/monthly_invoice_matrix.js
window.open_customer_invoice_dialog = function (docname) {
    if (!docname) return;

    frappe.db.get_doc("Customer Invoice", docname).then(doc => {
        let d = new frappe.ui.Dialog({
            title: __('Update Customer Invoice: {0}', [docname]),
            fields: [
                {
                    label: 'Status',
                    fieldname: 'status',
                    fieldtype: 'Select',
                    options: ['Draft', 'Unpaid', 'Partially Paid', 'Paid', 'Overdue'],
                    default: doc.status,
                    reqd: 1
                },
                {
                    label: 'Total Paid',
                    fieldname: 'invoice_paid_amount',
                    fieldtype: 'Currency',
                    default: doc.invoice_paid_amount,
                    reqd: 1
                }
            ],
            primary_action_label: __('Update Record'),
            primary_action(values) {
                frappe.call({
                    method: 'frappe.client.set_value',
                    args: {
                        doctype: 'Customer Invoice',
                        name: docname,
                        fieldname: values
                    },
                    freeze: true,
                    freeze_message: __('Saving...'),
                    callback: function () {
                        d.hide();
                        frappe.show_alert({ message: __('Invoice updated successfully'), indicator: 'green' });
                        frappe.query_report.refresh();
                    }
                });
            }
        });
        d.show();
    });
};

frappe.query_reports["Monthly Invoice Matrix"] = {
    "filters": [],
    "formatter": function (value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);

        if (!data) return value;

        if (column.fieldname === "customer_id" && !data.is_summary) {
            let color = data.overall_status === "Paid" ? "#155724" : "#856404";
            let background = data.overall_status === "Paid" ? "#d4edda" : "#fff3cd";
            return `<span style="background-color: ${background}; color: ${color}; padding: 8px 8px; border-radius: 4px; font-weight: 600; display: inline-block;">${data.customer_name || data.customer_id}</span>`;
        }

        let monthKey = column.fieldname;
        let docName = data[`doc_${monthKey}`];
        let status = data[`status_${monthKey}`];
        let invoiceDate = data[`date_${monthKey}`];

        if (docName && status !== "-") {
            let isPaid = status === "Paid";
            let background = isPaid ? "#d4edda" : "#fff3cd";
            let color = isPaid ? "#155724" : "#856404";
            return `<span onclick="open_customer_invoice_dialog('${docName}')"
                          style="background-color: ${background}; color: ${color}; padding: 8px 8px; border-radius: 4px; font-weight: 600; display: block; text-align: center; cursor: pointer;"
                          title="Click to quick edit invoice">
                          ${value}<br><small>${status} | ${invoiceDate || ''}</small>
                    </span>`;
        }

        return value;
    },
    "get_datatable_options": function (options) {
        return Object.assign(options, {
            checkboxColumn: false,
            cellHeight: 70,
        });
    }
};
